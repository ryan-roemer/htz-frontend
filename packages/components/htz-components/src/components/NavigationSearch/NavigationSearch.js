/* global document */
import React, { Fragment, } from 'react';
import { createComponent, FelaComponent, } from 'react-fela';
import { parseStyleProps, } from '@haaretz/htz-css-tools';

import IconClose from '../Icon/icons/IconClose';
import IconSearch from '../Icon/icons/IconSearch';
import Link from '../Link/Link';
import TextInput from '../TextInput/TextInput';

const wrapperStyle = ({ theme, miscStyles, }) => ({
  display: 'flex',
  extend: [
    ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
  ],
});

const myButtonStyle = ({ theme, isOpen, goSearch, }) => ({
  ...theme.type(-2),
  color: theme.color('neutral', '-3'),
  border: 'none',
  fontWeight: '700',
  minWidth: '6rem',
  padding: '1rem',
  ...(isOpen && {
    backgroundColor: theme.color('primary'),
    color: theme.color('neutral', '-10'),
  }),
  ...(goSearch && {
    backgroundColor: theme.color('primary', '-4'),
  }),
});
const MyButton = createComponent(myButtonStyle, 'button', [
  'onClick',
  'role',
  'aria-expanded',
]);

const Input = createComponent(({ theme, }) => ({
  flexGrow: '2',
  backgroundColor: theme.color('primary', '-4'),
}));

class NavigationSearch extends React.Component {
  state = {
    isOpen: false,
    query: null,
  };

  componentDidUpdate() {
    this.state.isOpen
      ? document.addEventListener('keydown', this.handleGlobalKeydown)
      : document.removeEventListener('keydown', this.handleGlobalKeydown);
  }

  changeState = () =>
    this.setState({
      isOpen: !this.state.isOpen,
    });

  focusOnInput = inputRef => inputRef && inputRef.focus();

  handleGlobalKeydown = e => {
    const key = e.which || e.keyCode;
    if (key === 27) {
      this.changeState();
    }
    else if (key === 13) {
      this.linkRef.click();
    }
  };

  recordQuery = e =>
    this.setState({
      query: e.target.value,
    });

  render() {
    // eslint-disable-next-line react/prop-types
    const { miscStyles, } = this.props;
    return (
      <FelaComponent
        miscStyles={miscStyles}
        rule={wrapperStyle}
        render={({
          className,
          theme: { navigationSearchI18n: { button, placeHolder, queryUrl, }, },
        }) => (
          <div className={className}>
            <MyButton
              isOpen={this.state.isOpen}
              onClick={this.changeState}
              role="button"
              aria-expanded={this.state.isOpen}
            >
              {this.state.isOpen ? (
                <IconClose size={3} color={[ 'neutral', '-10', ]} />
              ) : (
                <Fragment>
                  <IconSearch
                    size={3}
                    color="primary"
                    miscStyles={{
                      marginEnd: '2rem',
                    }}
                  />
                  <span>{button}</span>
                </Fragment>
              )}
            </MyButton>
            {this.state.isOpen && (
              <Fragment>
                <Input>
                  <TextInput
                    label={button}
                    labelHidden
                    placeholder={placeHolder}
                    type="text"
                    refFunc={this.focusOnInput}
                    onChange={this.recordQuery}
                    onBlur={this.changeState}
                    miscStyles={{
                      height: '100%',
                      marginTop: 'unset',
                      backgroundColor: 'transparent',
                      ':hover': {
                        backgroundColor: 'transparent',
                      },
                      ':focus-within': {
                        border: 'none',
                      },
                    }}
                  />
                </Input>
                <FelaComponent
                  rule={myButtonStyle}
                  goSearch
                  render={({ className, }) => (
                    <Link
                      // TODO: Change to Next link.
                      href={queryUrl(this.state.query) || '#'}
                      // eslint-disable-next-line no-return-assign
                      refFunc={linkRef => (this.linkRef = linkRef)}
                      className={className}
                      content={<IconSearch size={3} color="primary" />}
                    />
                  )}
                />
              </Fragment>
            )}
          </div>
        )}
      />
    );
  }
}

export default NavigationSearch;
