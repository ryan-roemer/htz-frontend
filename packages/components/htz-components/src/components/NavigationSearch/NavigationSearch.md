<!-- START doctoc generated TOC please keep comment here to allow auto update -->

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

**Table of Contents**

* [NavigationSearch examples](#NavigationSearch-examples)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### NavigationSearch examples

```jsx
<div style={{ direction: 'rtl' }}>
  <NavigationSearch />
</div>
```
