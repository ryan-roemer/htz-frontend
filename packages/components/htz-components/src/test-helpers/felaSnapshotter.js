import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
