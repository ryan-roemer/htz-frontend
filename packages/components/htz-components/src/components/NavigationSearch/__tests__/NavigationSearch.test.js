import React from 'react';
import felaSnapshotter from '../../../test-helpers/felaSnapshotter';
import NavigationSearch from '../NavigationSearch';

it('Navigation Search ', () => {
  const snapshot = felaSnapshotter(<NavigationSearch />);
  expect(snapshot).toMatchSnapshot();
});
