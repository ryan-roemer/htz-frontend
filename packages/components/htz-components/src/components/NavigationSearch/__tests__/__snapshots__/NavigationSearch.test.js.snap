// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Search  1`] = `
Object {
  "component": <div>
    <div
      className="htz-a"
    >
      <button
        aria-expanded={false}
        className="htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j"
        onClick={[Function]}
        role="button"
      >
        <svg
          className="htz-k htz-l htz-m htz-n htz-o"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
        >
           
          <path
            d="M247.6 228.5L175.2 156c11-15.2 17.5-33.8 17.5-54 0-50.8-41.3-92.1-92.1-92.1S8.5 51.3 8.5 102.1s41.3 92.1 92.1 92.1c20.6 0 39.7-6.8 55-18.3l72.3 72.3 19.7-19.7zm-147-57.2c-38.2 0-69.2-31.1-69.2-69.2s31-69.2 69.2-69.2 69.2 31.1 69.2 69.2-31.1 69.2-69.2 69.2z"
            fill="currentColor"
          />
        </svg>
        <span>
          חיפוש
        </span>
      </button>
    </div>
  </div>,
  "styles": ".htz-a{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex}.htz-f{color:#787878}.htz-g{border:none}.htz-h{font-weight:700}.htz-i{min-width:6rem}.htz-j{padding:1rem}.htz-k{vertical-align:-0.15em}.htz-l{color:#0B7EB5}.htz-m{fill:#FFF}.htz-n{font-size:3rem}.htz-o{margin-left:2rem}@media (max-width: 79.9375em){.htz-b{font-size:2rem}.htz-c{line-height:1.5em}}@media (min-width: 80em){.htz-d{font-size:2rem}.htz-e{line-height:1.5em}}",
}
`;
